python/ray/data/_internal/stream_split_dataset_iterator.py

```diff
@@ -0,0 +1,171 @@
+import copy
+import logging
+from typing import (
+    List,
+    Literal,
+    Optional,
+    Iterator,
+    Callable,
+    Any,
+    Union,
+    TYPE_CHECKING,
+)
+
+import ray
+
+from ray.data.dataset_iterator import DatasetIterator
+from ray.data.block import Block, DataBatch
+from ray.data.context import DatasetContext
+from ray.data._internal.execution.streaming_executor import StreamingExecutor
+from ray.data._internal.execution.legacy_compat import (
+    execute_to_legacy_bundle_iterator,
+)
+from ray.data._internal.block_batching import batch_block_refs
+from ray.data._internal.execution.operators.output_splitter import OutputSplitter
+from ray.data._internal.execution.interfaces import NodeIdStr
+from ray.types import ObjectRef
+from ray.util.scheduling_strategies import NodeAffinitySchedulingStrategy
+
+if TYPE_CHECKING:
+    import pyarrow
+    from ray.data import Dataset
+
+logger = logging.getLogger(__name__)
+
+
+class StreamSplitDatasetIterator(DatasetIterator):
+    """Implements a collection of iterators over a shared data stream."""
+
+    @staticmethod
+    def create(
+        base_dataset: "Dataset",
+        n: int,
+        equal: bool,
+        locality_hints: Optional[List[NodeIdStr]],
+    ) -> List["StreamSplitDatasetIterator"]:
+        """Create a split iterator from the given base Dataset and options.
+
+        See also: `Dataset.streaming_split`.
+        """
+        ctx = DatasetContext.get_current()
+        coord_actor = SplitCoordinator.options(
+            max_concurrency=n,
+            scheduling_strategy=NodeAffinitySchedulingStrategy(
+                ray.get_runtime_context().get_node_id(), soft=False
+            ),
+        ).remote(ctx, base_dataset, n, equal, locality_hints)
+        return [
+            StreamSplitDatasetIterator(base_dataset, coord_actor, i) for i in range(n)
+        ]
+
+    def __init__(
+        self,
+        base_dataset: "Dataset",
+        coord_actor: ray.actor.ActorHandle,
+        output_split_idx: int,
+    ):
+        self._base_dataset = base_dataset
+        self._coord_actor = coord_actor
+        self._output_split_idx = output_split_idx
+
+    def iter_batches(
+        self,
+        *,
+        prefetch_blocks: int = 0,
+        batch_size: int = 256,
+        batch_format: Literal["default", "numpy", "pandas"] = "default",
+        drop_last: bool = False,
+        local_shuffle_buffer_size: Optional[int] = None,
+        local_shuffle_seed: Optional[int] = None,
+        _collate_fn: Optional[Callable[[DataBatch], Any]] = None,
+    ) -> Iterator[DataBatch]:
+        """Implements DatasetIterator."""
+
+        def gen_blocks() -> Iterator[ObjectRef[Block]]:
+            future = self._coord_actor.get.remote(self._output_split_idx)
+            while True:
+                block = ray.get(future)
```

Comment:
Ah I missed that future is `ObjectRef[ObjectRef[Block]]`.

Reply:
Added explicit types here.